### ISSUE TITLE
WANTED: People to help maintain supercsv

### ISSUE THREAD

**Issue by kbilsted** [super-csv/super-csv#47]
Dear super-csv community, we are looking for people to help maintain super-csv. Time and motivation is at low, for both James and I. 

James has moved the project tremendously and I if only we can find people with as much enthusiasm and forward thinking as he, it would be great :) 

You don't need to be a supercsv expert to jump along, all you need is the time and motivation to do great stuff with a great library. There are plenty of issues to start solving, or if your clock it ticking in a different direction, raise new issues and lets see if we can agree that they are indeed good :)

**Comment by ChongTang** [comment]
Hi. I am a computer science Ph.D student. I am interested in maintaining SuperCSV. Maybe we can talk more?

**Comment by kbilsted** [comment]
Sounds great! There are plenty of issues here to start working on. Anything in particular you want to start improving?

**Comment by ChongTang** [comment]
@kbilsted I think it's better to start with fixing issues. So I can get familiar with everything through fixing issues. Do you have anything special or urgent that need to be fixed? If no, I will pick one of the issues to work on.

**Comment by kbilsted** [comment]
@ChongTang Do what you find the most fun..one interesting thing: how well is the support for Chinese source code. Maybe there are issues similar to #16 

Please raise issues before coding away on stuff not already reported so we can agree that it is a good idea 😋

Also issue #26 may be a good starting point

**Comment by ChongTang** [comment]
@kbilsted OK. I will take a look at these two issues first.

**Comment by kbilsted** [comment]
@ChongTang  I'm looking forward to seeing how you spell get and set in Chinese then. :)

**Comment by ZioberMichal** [comment]
Hi @kbilsted,
   I am a fun of this library and I will try to help you guys as much as I will able to. I have even started working on issue #16 but it is not easy to reproduce. I will try to figure out how to reproduce it and fix.

**Comment by kbilsted** [comment]
@ZioberMichal  That sounds fantastic. With you and maybe @ChongTang  on-board it is important to first discuss in an issue what you are working on so we don't all do the same work. Looking at #16 is a great start. Thanks!

**Comment by email2vimalraj** [comment]
I am in love with this library. I would like to contribute on any issue. I'm not an expert in java, but would take the issues which I can easily to start with my part of contribution to super-csv community. Thanks.

**Comment by kbilsted** [comment]
@email2vimalraj  great! just pick something from the issue list or add your own issues

**Comment by longtimeago** [comment]
@kbilsted I also don't mind to join if still needed :)

**Comment by kbilsted** [comment]
@longtimeago  I can't see why not ;) I see you already have  an idea in the open - I think its a great idea to expand the API that utilizes the new features of java.

**Comment by ash211** [comment]
I just submitted a PR at https://github.com/super-csv/super-csv/pull/103 and am happy to do a little tit-for-tat work on this project to get my PR merged.

Is there anything I can do for folks in exchange for getting code review on my PR from e.g. @kbilsted or @jamesbassett ?  I'm happy to do a ticket review to try to clear out old stuff, fill in some docs where people think they're lacking, or do a code review myself for someone else's outstanding pull request.  In general, help this project out as it's struggling for attention from the maintainers.

Any suggestions?

**Comment by kbilsted** [comment]
@ash211  sounds good. At the moment it is @ZioberMichal  who is doing all the heavy lifting. I'd gladly review your code changes and welcome you review others' PR's. Im sure we can all learn from each others points of views.

Are you asking for initial tasks to do ?

**Comment by ash211** [comment]
I've got one task in progress that I'm selfishly interested in seeing through to completion - #103 - and am offering to do another similarly-sized task as a "pay it forward" to the community.

**Comment by kbilsted** [comment]
@ash211  yeah thats how this project has always run - doing the things one think is fun to do ;)

**Comment by kbilsted** [comment]
@longtimeago  how are you doing? Find anything interesting to start poking at?

**Comment by longtimeago** [comment]
@kbilsted I'm fine, thanks! Just pretty busy on my main job now, sorry

**Comment by kbilsted** [comment]
@longtimeago  I know the feeling ;) Let us know when you (if ever) get more time on your hands ;) I'll delete the comments as to keep this thread down in size

**Comment by longtimeago** [comment]
@kbilsted thank you for understanding!

**Comment by wkrick** [comment]
@kbilsted I want to submit a new cellprocessor but GitHub is proving to be a very steep wall to climb just to submit 25 lines of code.

**Comment by kbilsted** [comment]
Just post it here then:)

**Comment by wkrick** [comment]
I've attached the cellprocessor.  Basically, it's a way to work around values that might have leading zeroes like ZIP codes.  It wraps the text with double quotes with an equals sign prepended like this...

="08109"

...this prevents the leading zeroes from getting lost when opening the CSV.  I'm not totally sure of the mechanics of how it works, but I think that it's forcing Excel to treat the field like a formula.

I tried the other "trick" of putting a single apostrophe at the beginning of the data to preserve the leading zeroes, but SuperCSV doesn't seem to allow doing that in a cellprocessor.  It seemed to keep adding a second "closing" apostrophe when I didn't want one.

More info on this leading zero issue...
https://social.msdn.microsoft.com/Forums/en-US/aae07b39-865f-4c68-a07f-7cad2dfd6733/how-do-i-open-csv-using-excel-without-deleting-leading-zeros?forum=isvvba


[PreserveLeadingZeroes.zip](https://github.com/super-csv/super-csv/files/592154/PreserveLeadingZeroes.zip)

**Comment by myselfandyi** [comment]
有没有操作csv文件的框架啊？

**Comment by SchlauFuchs** [comment]
If you like you can add me as contributor

**Comment by kbilsted** [comment]
Hi @tvand  not sure how we would go about it

**Comment by patric-r** [comment]
@tvand  https://commons.apache.org/proper/commons-csv/source-repository.html is broken

**Comment by tvand** [comment]
> @tvand https://commons.apache.org/proper/commons-csv/source-repository.html is broken (this is the link to "Source repository" in commons csv main menu)

See https://commons.apache.org/proper/commons-csv/scm.html

**Comment by tvand** [comment]
> Hi @tvand not sure how we would go about it

If you were interested, I could go investigate. AFAIK, an Intellectual property clearance would be needed. See http://incubator.apache.org/ip-clearance/ on how something like this looks.

Bye, Thomas.

**Comment by kbilsted** [comment]
But the code is already under apache license. And they have another csv implementation incompatible with this one.

**Comment by tvand** [comment]
Yes and that probably makes it easier. We would never "just take" your code, even if the license said we could. We would rather have you donate it. The ASF prefers to have your written consent. I would offer to organize most of the required steps, if you want.

Bye, Thomas.

**Comment by kbilsted** [comment]
@tvand  who can contribute that are prevented from contributing today? And why wouldn't they want to contribute today already?

**Comment by tvand** [comment]
> @tvand who can contribute that are prevented from contributing today? And why wouldn't they want to contribute today already?

I'm just hoping that people already contributing to commons-csv or using it will be attracted by the significantly more sophisticated features of super-csv - under the umbrella of Apache. I cannot offer more than this hope at this time.

Bye, Thomas.

**Comment by moarcode** [comment]
@kbilsted are you still looking for contributors?

**Comment by kbilsted** [comment]
@moarcode  yes that would be great. But I don't see you doing any public work for the past 3 years on GH?

**Comment by dota17** [comment]
Hi, @kbilsted 
Are you still looking for a maintainer? I'm interested in it.

**Comment by kbilsted** [comment]
@dota17  sounds great!

**Comment by SchlauFuchs** [comment]
I could help, too. What is on the backlog?

**Comment by kbilsted** [comment]
@SchlauFuchs fantastic . We need to have fun. So whatever tickles you... 

Also We need to deal with issues, review the pr's and make ready for a v2.50 release before long

**Comment by dota17** [comment]
Hi, @kbilsted 
What features and issues are more urgent for version 2.5.0? I can prioritize. And what is your plan for 2.5.0?

**Comment by kbilsted** [comment]
@SchlauFuchs  found anything of interest?

**Comment by SchlauFuchs** [comment]
The first thing I found (looking for the help wanted tag) was the one about
empty rows made of commas. Someone has done work on it already, I wonder
why it wasn't merged yet.

I did not look into the code, only the backlog.

I could tackle this

**Comment by SchlauFuchs** [comment]
Hi,

Kasper, can you review/merge my PR at
https://github.com/super-csv/super-csv/pull/196 ?

I had a quick a quick review of some sources, especially wondering that the
java classlevel of the project is still 5. I had to dig out a JDK8 from
deep in the web to be able to build this under such settings. I really
recommend to upgrade the project to at least classlevel 8 for compatibility
reasons, but that is already so much EOL that going to 11 would make sense.
I start a branch in my fork for that and give the project the
necessary changes.

Cheers,
Kai

**Comment by brbog** [comment]
I just reviewed https://github.com/super-csv/super-csv/pull/216 (all 7 commits).

There seem to be people wanting to contribute to this project, but even low risk pull requests don't get approved. While I'm not volunteering to take this project in any special direction, I wouldn't mind reviewing low risk stuff that doesn't change the architecture so at least basic maintenance would still be happening (bugfixes & staying up to date with versions of libraries or Java).

It would be a pitty to see this project end up dead. It's probably still the best library available, even if it doesn't have annotation bells and whistles.

**Comment by patric-r** [comment]
As the maintainer(s) still seems to be unresponsive, we should think about forking this project, apply the PR/fixes there and finally release a new version.

**Comment by brbog** [comment]
@patric-r , cc: @kbilsted :
I second that, but seeing the amount of issues and pull requests, I was wondering if it's even possible to copy all that good input to a fork. It would be easier to first get things cleaned up on this fork and then pass it on to another maintainer. Maybe others can get more rights on this fork from the original maintainers first?

**Comment by patric-r** [comment]
@brbog If you think the PRs will get merged soon, I agree.
But to be honest and when considerung the previous years - I doubt that.
Have a look at this issue. The original author asked 2015 for a handover but nothing really happened so far even though there were several promising volunteers. Last action from @kbilsted had been in 2020 here...

Therefore I vote for what will happen sooner or later anyhow (if SuperCSV has a future at all):
A community-drive active fork needs to get created.
Maybe a deadline should be issued to the original author beforehand.

PRs from the original project can still get merged...

**Comment by brbog** [comment]
@patric-r : I'm afraid you are right and @kbilsted probably has a mail filter to ignore all messages originating from super-csv (addressing him is a bit of a test for this actually).

Maybe people can answer on this thread:
- if they are willing to help get everything code reviewed/fixed into a new fork
- if they know of someone who deserves the maintainer-role (that would become the account where the fork is maintained)

A little effort (2h or so) from multiple people could help a lot (it's a nice straightforward library), I don't mind helping out a little.

**Comment by ash211** [comment]
I would propose that whoever interested folks go about forking via:

- pick a new name for this project that's a slight variant, like `super-csv2`
- create the github project+repo for it, like github.com/super-csv2/super-csv2
- push this code over and do a first release with the same version number of the last super-csv release: [2.4.0](https://github.com/super-csv/super-csv/releases)
  - this way super-csv 2.4.0 and super-csv2 2.4.0 are identical code (helps people with migration)
- port PRs from super-csv over to super-csv2 and carry the project on in whatever way the new contributors desire

**Comment by patric-r** [comment]
Sounds good. But I still think we should offer @kbilsted a deadline to help the community by either adding people as maintainers to this repo (preferred) or at least accept the most important PRs.

If he doesn't react, we should go for a fork.

@SchlauFuchs @dota17 What do you think?

**Comment by ZioberMichal** [comment]
Hey, I had a permission to the organisation but it looks like it was removed and now I can not add anyone as a maintainer.

**Comment by ZioberMichal** [comment]
Also, I can try to accept most important MR. Please list them and help me to approve. We can try to release version 2.5.0 together.

WDYT, @SchlauFuchs , @dota17 , @patric-r , @ash211

**Comment by ZioberMichal** [comment]
I tried to approve MR created by dependency bot but it is still blocked so it means my collaborator role is not enough to push it further.

**Comment by kbilsted** [comment]
Many thanks to all the people who recently showed interest in contributing. It's great to see such energy ;-)

I am very interested in moving the project onwards. As many of you know, the package is fairly wide-spread so the last thing we want is to some troll taking over and committing malware into the package. And there has been struggles with people signaling interest in the project only to disappear unfortunately.

Other struggles in the past with PR's is that people simply code before discussing the validity of their ideas and/refusing to implement tests that show the new feature is actually working. Working tests significantly reduces the work code reviews entails. Finally, the importance of performance in the library is a topic not all grasp.

I hope all you future maintainers share these points of views. If not please voice them now, perhaps there are better values we should cultivate?

Personally, I'd like to participate in discussions and reviews, but realistically won't be coding much myself. If someone will help me setup the tool-chain I can do the releases perhaps in conjunction with project veterans. That be for starters. Naturally more people should gain access when we know each other a little better.

**Comment by patric-r** [comment]
Hej @kbilsted,  great to see you alive here again ;)

While it cannot be completely ruled out, I don't see big risks about "malware commiters". 
This rarely happens. The risk supercsv is slowly going to die if you do nothing is much higher :)

Of course you should choose wisely who will gain commiter privileges but if you do this for people which are active on github for a long time and especially contributing to this project for years, the risk is rather theoretical for me.
And at the end, giving others commiter access doesn't mean that you're out of the game.

Let's face the reality:
Either you're going to still keep the keys (which is my favorite option) - but then should react on the most important PRs at least (and at the end, ideally merging them) and get a 2.5.0 release out as soon as possible.
Or you hand some of your keys over to other volunteers and see where things are going.
I strongly believe that there are many other clever people here who will respect your architectural and quality concerns. 

If neither of the options work for you, we should continue with a fork (hopefully there is still enough momentum left for that).
It would be a pitty if we would need to move on to another csv library...


@ash211 @brbog  @ZioberMichal  What do you think?

**Comment by ZioberMichal** [comment]
I agree with @patric-r. Let's release version 2.5.0 which is almost ready, add some version upticks for libraries suggested by dependency bot, and maybe some small MR and release it.

After that we can discuss who can lead the project after @kbilsted. Some time ago I thought I will be able to do that but I could not find time for it. Nothing changed since then. Right now, I can only play a support role.

**Comment by brbog** [comment]
@patric-r : well worded.

@kbilsted (and others)
Current reality in Belgium is that most people will conservatively select the Apache Commons CSV library or look at articles like https://www.baeldung.com/java-csv in which supercsv is not mentioned. Anyone else will probably be looking for something that satisfies their "latest-syntax-hype-fetish".

The weakest spot of supercsv is that it looks abandoned/unmaintained, so it has become nearly impossible to suggest using it in a professional environment. Most people are sensible enough to avoid doing crazy things with csv, so using other libs with fingers crossed and "fix problems only if they appear in practice" actually works, but I rather like a library that can handle every weird use-case imaginable while still being easy to use :-). If it would get minor updates once in a while (Java 8 is reaching end of life), that would be a great marketing boost.

@kbilsted : If you are afraid of getting a stained legacy, you could consider moving to a 3.0.0 release so you can easily demarcate your involvement until the last 2.x version (weirder things have happened with version numbers, see http://www.slackware.com/faq/do_faq.php?faq=general#0 ).

I'm hoping that we can get supercsv up to date with minimal effort. The library was great from the beginning and anyone who believes in a fundamentally different design will have wandered off to a different library already I think :-).

**Comment by jjwestwood** [comment]
Hi all in the super-csv community, I see the last activity here was quite some time ago, as the comment above states from the outside super-csv looks abandoned/unmaintained.

I'm in the process of starting a new project and super-csv has been the default choice in the past, it's proven to be stable and is well known, however I'm struggling with how little movement there has been in so long. 

I'm just reaching out to see if there is someone closer to the project who could provide some information on if this is still a viable library or if there there are any plans/potential to move this forward in the future?

Kind regards

Jon
